sql/src/main/java/io/crate/planner/operators/InsertFromValues.java

```diff
@@ -0,0 +1,766 @@
+import io.crate.planner.ExecutionPlan;
+import io.crate.planner.PlannerContext;
+import io.crate.types.DataType;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.create.CreatePartitionsRequest;
+import org.elasticsearch.action.admin.indices.create.TransportCreatePartitionsAction;
+import org.elasticsearch.action.bulk.BackoffPolicy;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.routing.ShardIterator;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper;
+import org.elasticsearch.index.IndexNotFoundException;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.StreamSupport;
+
+import static io.crate.data.SentinelRow.SENTINEL;
+import static io.crate.execution.engine.indexing.ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING;
+
+
+public class InsertFromValues implements LogicalPlan {
+
+    private static final BackoffPolicy BACK_OFF_POLICY = LimitedExponentialBackoff.limitedExponential(1000);
+
+    private final TableFunctionRelation tableFunctionRelation;
+    private final ColumnIndexWriterProjection writerProjection;
+
+    InsertFromValues(TableFunctionRelation tableFunctionRelation,
+                     ColumnIndexWriterProjection writerProjection) {
+        this.tableFunctionRelation = tableFunctionRelation;
+        this.writerProjection = writerProjection;
+    }
+
+    @Override
+    public StatementType type() {
+        return StatementType.INSERT;
+    }
+
+    @Override
+    public void execute(DependencyCarrier dependencies,
+                        PlannerContext plannerContext,
+                        RowConsumer consumer,
+                        Row params,
+                        SubQueryResults subQueryResults) {
+        DocTableInfo tableInfo = dependencies
+            .schemas()
+            .getTableInfo(
+                RelationName.fromIndexName(writerProjection.tableIdent().fqn()),
+                Operation.INSERT);
+
+        InputFactory inputFactory = new InputFactory(dependencies.functions());
+        InputFactory.Context<CollectExpression<Row, ?>> context =
+            inputFactory.ctxForInputColumns(plannerContext.transactionContext());
+
+        var allColumnSymbols = InputColumns.create(
```

Comment:
Maybe we could add a visual representation of the different symbols as comment, so that it becomes a bit clearer how they relate to each other.

Reply:
I think this would still be helpful to improve the readability.

Reply:
will do that 👍

Reply:
added some kind of diagram ) https://github.com/crate/crate/pull/9465/files#diff-a74268cd5e9c526d14e66b3dc5daa81aR152